File: SampleProjects/GoT-Characters-master/Objective-C/GoT/Configurator.m
//
//  Configurator.m
//  GoT
//
//  Created by Paciej on 25/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

#import "Configurator.h"
#import "AsyncLoadConfiguration.h"
#import "Loader.h"
#import "FavouriteTableViewCell.h"
#import "NSData+JSON.h"

static NSString *kItems = @"items";
static NSString *kId = @"id";
static NSString *kTitle = @"title";
static NSString *kAbstract = @"abstract";
static NSString *kThumbnailURL = @"thumbnail";
static NSString *kArticleBaseURL = @"basepath";
static NSString *kArticleURL = @"url";

static NSString *kCellReuseIdentifier = @"MainViewControllerCell";

@implementation Configurator

+ (nonnull MainViewController *)configuredMainViewController {
    AsyncLoadConfiguration *asyncLoadConfiguration =
    [Configurator configuredCharactersAsyncLoadConfiguration];
    Loader *loader = [Configurator configuredLoader];
    ArticlesRepository *repository = [Configurator configuredArticlesRepository];
    DataSource *dataSource = [Configurator configuredArticlesDataSource];
    
    return [[MainViewController alloc] initWithLoader:loader
                               asyncLoadConfiguration:asyncLoadConfiguration
                                           dataSource:dataSource
                                   articlesRepository:repository];
}

+ (nonnull DetailsViewController *)
configuredDetailsViewControllerWithArticle:(nonnull Article *)article
favourite:(BOOL)favourite {
    return [[DetailsViewController alloc] initWithArticle:article
                                                favourite:favourite];
}

+ (nonnull ArticlesRepository *)configuredArticlesRepository {
    return [[ArticlesRepository alloc]
            initWithDefaults:[NSUserDefaults standardUserDefaults]];
}

+ (nonnull DataSource *)configuredArticlesDataSource {
    return [[DataSource alloc] initWithCellConfigureBlock:nil
                                      cellReuseIdentifier:kCellReuseIdentifier];
}

+ (nonnull Loader *)configuredLoader {
    NSURLSession *session = [NSURLSession
                             sessionWithConfiguration:[NSURLSessionConfiguration
                                                       defaultSessionConfiguration]];
    return [[Loader alloc] initWithWebserviceURLString:@"" session:session];
}

+ (nonnull AsyncLoadConfiguration *)configuredCharactersAsyncLoadConfiguration {
    return [[AsyncLoadConfiguration alloc]
            initWithResponseParsingBlock:^id(NSData *result) {
                id JSON = result.JSONObject;
                if (!JSON) {
                    NSLog(@"loadAsynchronously error: cannot create JSON object from "
                          @"server response.");
                    return nil;
                }
                NSString *basePath = [JSON objectForKey:kArticleBaseURL];
                NSArray *items = [JSON objectForKey:kItems];
                NSMutableArray *articles =
                [[NSMutableArray alloc] initWithCapacity:items.count];
                for (NSDictionary *item in items) {
                    @autoreleasepool {
                        Article *article = [[Article alloc]
                                            initWithIdentifier:item[kId]
                                            title:item[kTitle]
                                            abstract:item[kAbstract]
                                            urlString:[NSString stringWithFormat:@"%@%@", basePath,
                                                       item[kArticleURL]]
                                            thumbnailURLString:item[kThumbnailURL]];
                        [articles addObject:article];
                    }
                }
                return articles;
            } webserviceEndpoint:@"http://gameofthrones.wikia.com/api/v1/Articles/Top"
            webserviceQuery:@"?expand=1&category=Characters&limit=75"];
}

@end

File: SampleProjects/GoT-Characters-master/Swift/GoT/Configurator.swift
//
//  Configurator.swift
//  GoT
//
//  Created by Paciej on 25/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

import Foundation

private var kItems = "items"
private var kId = "id"
private var kTitle = "title"
private var kAbstract = "abstract"
private var kThumbnailURL = "thumbnail"
private var kArticleBaseURL = "basepath"
private var kArticleURL = "url"
private var kCellReuseIdentifier = "MainViewControllerCell"

class Configurator: NSObject {
    class func configuredMainViewController() -> MainViewController {
        let asyncLoadConfiguration = Configurator.configuredCharactersAsyncLoadConfiguration()
        let loader = Configurator.configuredLoader()
        let repository = Configurator.configuredArticlesRepository()
        let dataSource = Configurator.configuredArticlesDataSource()

        return MainViewController(loader: loader, asyncLoadConfiguration: asyncLoadConfiguration, dataSource: dataSource, articlesRepository: repository)
    }

    class func configuredDetailsViewController(with article: Article, favourite: Bool) -> DetailsViewController {
        return DetailsViewController(article: article, favourite: favourite)
    }

    class func configuredArticlesRepository() -> ArticlesRepository {
        return ArticlesRepository(defaults: UserDefaults.standard)
    }

    class func configuredArticlesDataSource() -> DataSource {
        return DataSource(cellConfigureBlock: nil, cellReuseIdentifier: kCellReuseIdentifier)
    }

    class func configuredLoader() -> Loader {
        let session = URLSession(configuration: URLSessionConfiguration.default)
        return Loader(webserviceURLString: "", session: session)
    }

    class func configuredCharactersAsyncLoadConfiguration() -> AsyncLoadConfiguration {
        return AsyncLoadConfiguration(responseParsingBlock: { result in
            let JSON = result.jsonObject
            if JSON == nil {
                print("""
                loadAsynchronously error: cannot create JSON object from \
                server response.
                """)
                return nil
            }
            let basePath = (JSON as? [AnyHashable : Any])?[kArticleBaseURL] as? String
            let items = (JSON as? [AnyHashable : Any])?[kItems] as? [Any]
            var articles = [AnyHashable](repeating: 0, count: items?.count ?? 0)
            for item: [AnyHashable : Any]? in items as? [[AnyHashable : Any]?] ?? [] {
                autoreleasepool {
                    let article = Article(identifier: item?[kId] as? String ?? "", title: item?[kTitle] as? String ?? "", abstract: item?[kAbstract] as? String ?? "", urlString: "\(basePath ?? "")\(item?[kArticleURL])", thumbnailURLString: item?[kThumbnailURL] as? String ?? "") as? Article
                    if let anArticle = article {
                        articles.append(anArticle)
                    }
                }
            }
            return articles
        }, webserviceEndpoint: "http://gameofthrones.wikia.com/api/v1/Articles/Top", webserviceQuery: "?expand=1&category=Characters&limit=75")
    }
}